File: swift/Sources/MudbaseShowcaseEcommerce/Support/JSONField.swift
import Foundation

/// Mudbase Collection fields have a fixed type enum (string, number, boolean, date, email, url,
/// enum, reference) with no native array/object type. Anything shaped like a list or a nested
/// record (order line items, a shipping address, extra product images) is stored as a JSON string
/// in a `string` field and parsed at the edges — a real, documented constraint of the platform's
/// Collections feature, mirroring `web/src/lib/json-field.ts` exactly.
enum JSONField {
    static func parse<T: Decodable>(_ value: String?, as type: T.Type, fallback: T) -> T {
        guard let value, let data = value.data(using: .utf8) else { return fallback }
        let decoder = JSONDecoder()
        return (try? decoder.decode(T.self, from: data)) ?? fallback
    }

    static func stringify(_ value: some Encodable) -> String {
        let encoder = JSONEncoder()
        guard let data = try? encoder.encode(value), let string = String(data: data, encoding: .utf8) else {
            return "[]"
        }
        return string
    }
}

File: ruby/lib/mudbase/json_field.rb
# frozen_string_literal: true

require "json"

module Mudbase
  # Mudbase Collection fields have a fixed type enum (string, number, boolean, date, email,
  # url, enum, reference) with no native array/object type. Anything shaped like a list or a
  # nested record (order line items, a shipping address, extra product images) is stored as a
  # JSON string in a `string` field and parsed at the edges - a real, documented constraint of
  # the platform's Collections feature, not a workaround specific to this app. Mirrors
  # `web/src/lib/json-field.ts` in the reference implementation.
  module JsonField
    # @param value [String, nil]
    # @param fallback [Object] returned as-is if `value` is blank or invalid JSON.
    # @return [Object] parsed with symbol keys, matching every other Hash in this app (see
    #   ClientFactory::OBJECT_RESPONSE) - callers read `item[:productId]`, `address[:fullName]`,
    #   never string keys.
    def self.parse(value, fallback)
      return fallback if value.nil? || value.empty?

      JSON.parse(value, symbolize_names: true)
    rescue JSON::ParserError
      fallback
    end

    def self.stringify(value)
      value.to_json
    end
  end
end
